### PR TITLE
Add Orchard and Unified Address support to z_viewtransaction 

### LINKED ISSUE #5186: z_viewtransaction: Figure out how to render UAs
Where possible, it would be great to render UAs inside `z_viewtransaction` output that matches what users might actually have used for those transactions. What this means is:

- If the user spent funds from a UA, show that UA in the output.
  - With domain separation of receivers, we should be able to link spent funds to specific Unified accounts.
  - Show the default UA for that account.
- If the user received funds to a UA, show that UA in the output.
  - We should be able to derive the exact UA that was used (e.g. from the diversifier in the note, recover the diversifier index, and use this to find or re-derive the UA).
  - This includes moving funds around inside the wallet with e.g. `z_mergetoaddress`.
- If the user sent funds to an external UA, show that UA in the output.
  - This would probably only be possible if we cache the recipient in `z_sendmany` (or other APIs).
  - For a recovery from seed phrase, we could only show a single-receiver UA (or for Sapling and transparent, a legacy format address).

### PR DESCRIPTION
**Pull request by nuttycom**
Closes #5186.

### CODE REVIEW

File: src/zcash/Address.hpp [hunk 1]
```diff
@@ -67,6 +67,16 @@ class UnifiedAddress {
 public:
     UnifiedAddress() {}
 
+    static std::optional<UnifiedAddress> ForReceivers(const std::vector<Receiver>& receivers) {
+        UnifiedAddress result;
+        for (const auto& receiver : receivers) {
+            if (!result.AddReceiver(receiver)) {
```
Comment (str4d) [inline src/zcash/Address.hpp:73]: This is adding another use of `UnifiedAddress::AddReceiver`, which requires enforcing an assumption about receiver order that is not being enforced here, and which I do not want to risk us accidentally breaking before we fix #5456. Replace this with a `UnifiedAddress::SingleReceiver` method that takes a `Receiver receiver` and returns a `std::optional<UnifiedAddress>`.

File: src/wallet/walletdb.cpp [hunk 2]
```diff
@@ -892,6 +892,8 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
                 strErr = "Error in wallet database: recipientmapping UA does not contain recipient";
                 return false;
             }
+
+            pwallet->LoadRecipientMapping(txid, RecipientMapping(ua.value(), recipient));
```
Comment (str4d) [inline src/wallet/walletdb.cpp:896]: This is silently tucked away in an unrelated commit. Move this into a separate commit with an explanation for this change in the commit message.

File: src/wallet/wallet.h [hunk 3]
```diff
@@ -1200,6 +1200,8 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
         pwalletdbEncryption = NULL;
     }
 
+    OrchardWallet& GetOrchardWallet() { return orchardWallet; }
```
Comment (str4d) [inline src/wallet/wallet.h:1203]: Do we really want to expose `OrchardWallet` outside of `CWallet`, rather than it being an internal implementation detail?

File: src/wallet/wallet.h [hunk 4]
```diff
@@ -1272,6 +1274,8 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
 
     std::map<uint256, CWalletTx> mapWallet;
 
+    std::map<uint256, std::vector<RecipientMapping>> sendRecipients;
```
Comment (str4d) [inline src/wallet/wallet.h:1277]: This should go in the separate commit with the other unexplained changes (like the addition of `LoadRecipientMapping`).

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -853,6 +853,67 @@ bool CWallet::LoadUnifiedAddressMetadata(const ZcashdUnifiedAddressMetadata &add
                 addrmeta.GetReceiverTypes());
 }
 
+PaymentAddress CWallet::GetPaymentAddressForRecipient(const uint256& txid, const libzcash::RecipientAddress recipient) const {
+    auto recipientsPtr = sendRecipients.find(txid);
+
+    // TODO: We should scrub change addresses from the results here.
```
Comment (str4d) [inline src/wallet/wallet.cpp:859]: Open an issue for this, blocking at a minimum 5.0.0, and ideally 4.7.0. Change addresses must not ever be shown to users, per our previous discussions about UX; if we ever show a change address to a user, we have to assume it can receive external funds, which breaks a bunch of the UX properties we're after. We can _maybe_ get away without anyone noticing or starting to rely on this in 4.7.0, but if it is the behaviour of 5.0.0 then we're screwed from a UX perspective.

What we need to do is replace any change address in the UX with a corresponding external address (a sender UA involved in the transaction that we can identify as coming from the same UFVK, or the default account address as a fallback).

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -853,6 +853,67 @@ bool CWallet::LoadUnifiedAddressMetadata(const ZcashdUnifiedAddressMetadata &add
...
+    auto defaultAddress = [&]() -> PaymentAddress {
```
Comment (str4d) [inline src/wallet/wallet.cpp:860]: This doesn't currently implement the policy we decided on and documented in #5575. At a minimum we should implement the non-contextual baseline policy:
> If the receiver matches a UFVK in the wallet, and we have no information about when this receiver may have been first used, show the UA corresponding to the most recent receiver types policy that includes the receiver's type.

Then add a TODO referencing the issue for the remaining contextual work.

File: src/wallet/rpcwallet.cpp [hunk 7]
```diff
@@ -4267,7 +4282,57 @@ UniValue z_viewtransaction(const UniValue& params, bool fHelp)
...
+        OrchardActionSpend orchardActionSpend = pair.second;
+        auto outpoint = orchardActionSpend.GetOutPoint();
+        auto receivedAt = orchardActionSpend.GetReceivedAt();
+        auto noteValue = orchardActionSpend.GetNoteValue();
+        // if mapWallet doesn't contain the spent outpoint's txid,
+        // the wallet is corrupt, so using `.at` here is fine.
+
+        UniValue entry(UniValue::VOBJ);
+        entry.pushKV("type", ADDR_TYPE_ORCHARD);
+        entry.pushKV("spend", (int) actionIdx);
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:4302]: The help text labels this `actionspend`, but I think `action` would be better (or whatever we use for these elsewhere).
```suggestion
        entry.pushKV("action", (int) actionIdx);
```
In any case, make this consistent with the help text.

File: src/wallet/rpcwallet.cpp [hunk 8]
```diff
@@ -4267,7 +4282,57 @@ UniValue z_viewtransaction(const UniValue& params, bool fHelp)
...
+        entry.pushKV("txidPrev", outpoint.hash.GetHex());
+        entry.pushKV("outputPrev", (int) outpoint.n);
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:4304]: The help text uses:
```suggestion
        entry.pushKV("actionPrev", (int) outpoint.n);
```

File: src/wallet/rpcwallet.cpp [hunk 9]
```diff
@@ -4267,7 +4282,57 @@ UniValue z_viewtransaction(const UniValue& params, bool fHelp)
...
+        auto ua = pwalletMain->FindUnifiedAddressByReceiver(receivedAt);
+        assert(ua.has_value());
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:4306]: I assume this is because we can't import UFVKs at present? Add a TODO referencing the issue about adding UFVK imports.

File: src/wallet/rpcwallet.cpp [hunk 10]
```diff
@@ -4267,7 +4282,57 @@ UniValue z_viewtransaction(const UniValue& params, bool fHelp)
...
+
+        // Show the address that was cached at transaction construction as the
+        // recipient.
+        std::string address = keyIO.EncodePaymentAddress(
+            pwalletMain->GetPaymentAddressForRecipient(txid, recipient)
+        );
+
+        UniValue entry(UniValue::VOBJ);
+        entry.pushKV("type", ADDR_TYPE_ORCHARD);
+        entry.pushKV("output", (int) actionIdx);
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:4328]: The help text calls this `actionoutput`, but again I think this should be `action` for consistency (since it's referring to an index into `vActionsOrchard`):
```suggestion
        entry.pushKV("action", (int) actionIdx);
```
Update the help text to be consistent.

File: src/wallet/rpcwallet.cpp [hunk 11]
```diff
@@ -4267,7 +4282,57 @@ UniValue z_viewtransaction(const UniValue& params, bool fHelp)
...
+        entry.pushKV("outgoing", orchardActionOutput.IsOutgoing());
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:4329]: Update the help text to indicate this will be present for Orchard as well as Sapling.

File: qa/rpc-tests/wallet_listreceived.py [hunk 12]
```diff
@@ -435,9 +442,101 @@ def test_received_sapling(self, height):
...
+        assert_equal(outputs[1]['output'], 1)
+        assert_equal(outputs[1]['outgoing'], False)
+        assert_equal(outputs[1]['memo'], no_memo)
+        assert 'memoStr' not in outputs[1]
+
+        self.generate_and_sync(height+3)
+
+        opid = self.nodes[1].z_sendmany(uao, [
+            {'address': uaso, 'amount': Decimal('0.4')}
+        ])
```
Comment (str4d) [inline qa/rpc-tests/wallet_listreceived.py:501]: Add a recipient address from a different node, so we can confirm we see `outgoing: true`.

File: qa/rpc-tests/wallet_listreceived.py [hunk 13]
```diff
@@ -435,9 +442,101 @@ def test_received_sapling(self, height):
...
+        assert_equal(outputs[0]['type'], 'orchard')
+        assert_equal(outputs[0]['address'], uaso)
+        assert_equal(outputs[0]['value'], Decimal('0.4'))
+        assert_equal(outputs[0]['valueZat'], 40000000)
+        assert_equal(outputs[0]['outgoing'], False)
+        assert_equal(outputs[0]['memo'], no_memo)
+
+        assert_equal(outputs[1]['type'], 'orchard')
+        # TODO: scrub change addresses
+        #assert_equal(outputs[1]['address'], '<change>')
```
Comment (str4d) [inline qa/rpc-tests/wallet_listreceived.py:530]: Oh hmm, I guess that's another approach: just don't show an address. If we do this, I'd be inclined to have a `change: true` field and just omit the `address` field entirely (so callers don't have to special-case the field to parse the special string `<change>`).

**Review by str4d — CHANGES_REQUESTED** [review]
Generally LGTM

File: src/wallet/wallet.h (hunk 4 above)
Comment (nuttycom) [reply src/wallet/wallet.h:1277]: This is necessary to be able to look up recipient UAs by receiver. This had been left incomplete in #5549, and I hadn't realized it.

File: src/wallet/wallet.cpp (hunk 5 above)
Comment (nuttycom) [reply src/wallet/wallet.cpp]: This has now been fixed.

File: qa/rpc-tests/wallet_accounts.py [hunk 14]
```diff
@@ -173,11 +173,10 @@ def run_test(self):
         txid = wait_and_assert_operationid_status(self.nodes[0], opid)
 
         # The wallet should detect the new note as belonging to the UA.
-        # TODO: Uncomment once z_viewtransaction shows Orchard details (#5186).
-        #tx_details = self.nodes[0].z_viewtransaction(txid)
-        #assert_equal(len(tx_details['outputs']), 1)
-        #assert_equal(tx_details['outputs'][0]['type'], 'orchard')
-        #assert_equal(tx_details['outputs'][0]['address'], ua0)
+        tx_details = self.nodes[0].z_viewtransaction(txid)
```
Comment (str4d) [inline qa/rpc-tests/wallet_accounts.py:176]: There are more test lines to uncomment below.

File: src/wallet/wallet.cpp (hunk 6 above)
Comment (nuttycom) [reply src/wallet/wallet.cpp:865]: #5706

File: qa/rpc-tests/wallet_accounts.py (hunk 14 above)
Comment (nuttycom) [reply qa/rpc-tests/wallet_accounts.py:176]: If CI passes, I'll submit that as a separate PR.

File: src/wallet/rpcwallet.cpp (hunk 11 above)
Comment (nuttycom) [reply src/wallet/rpcwallet.cpp:4350]: I'll also add this to a subsequent PR so as not to block CI.

Comment (nuttycom) [reply src/wallet/rpcwallet.cpp:4350]: (The additional uncommented tests do pass for me locally.)

File: src/keystore.cpp [hunk 15]
```diff
@@ -398,33 +398,36 @@ CBasicKeyStore::GetUFVKMetadataForAddress(const libzcash::UnifiedAddress& addr)
...
+            } else if (rmeta.has_value()) {
+                ufvkId = rmeta.value().GetUFVKId();
+                j = rmeta.value().GetDiversifierIndex();
             }
         }
     }
 
     if (ufvkId.has_value()) {
-        return std::make_pair(ufvkId.value(), j);
+        return AddressUFVKMetadata(ufvkId.value(), j, false);
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/keystore.cpp:430]: See the assertion: https://github.com/zcash/zcash/pull/5700/files/a1bd440f2b0be351eeeffd40230441e3363a5164#diff-327bdf36f54a34375e6dd162366d1cdcf1ae6ec12d8097abe90576e5a695ec20R403

File: src/wallet/wallet.cpp [hunk 16]
```diff
@@ -1486,7 +1607,7 @@ set<uint256> CWallet::GetConflicts(const uint256& txid) const
 
     for (uint32_t i = 0; i < wtx.GetOrchardBundle().GetNumActions(); i++) {
         OrchardOutPoint op(wtx.GetHash(), i);
-        auto potential_spends = pwalletMain->orchardWallet.GetPotentialSpends(op);
+        auto potential_spends = orchardWallet.GetPotentialSpends(op);
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/wallet.cpp:1610]: We only have one global pwalletMain, but yeah, this is not what one would want.

File: src/wallet/wallet.cpp [hunk 17]
```diff
@@ -853,6 +857,123 @@ bool CWallet::LoadUnifiedAddressMetadata(const ZcashdUnifiedAddressMetadata &add
...
+                        }
+                    }
+                }
+
+                return libzcash::PaymentAddress{UnifiedAddress::ForSingleReceiver(addr)};
+            }
+        }, recipient);
+    };
+
+    auto recipientsPtr = sendRecipients.find(txid);
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/wallet.cpp:920]: It's only populated for transaction recipients.

File: src/keystore.cpp [hunk 18]
```diff
@@ -465,30 +468,35 @@ std::optional<libzcash::UFVKId> CBasicKeyStore::GetUFVKIdForViewingKey(const lib
...
+std::optional<AddressUFVKMetadata> FindUFVKId::operator()(const libzcash::SaplingPaymentAddress& saplingAddr) const {
     const auto saplingIvk = keystore.mapSaplingIncomingViewingKeys.find(saplingAddr);
     if (saplingIvk != keystore.mapSaplingIncomingViewingKeys.end()) {
         // We have either generated this as a receiver via `z_getaddressforaccount` or a
         // legacy Sapling address via `z_getnewaddress`, or we have previously detected
         // this via trial-decryption of a note.
         const auto ufvkId = keystore.mapSaplingKeyUnified.find(saplingIvk->second);
         if (ufvkId != keystore.mapSaplingKeyUnified.end()) {
-            return std::make_pair(ufvkId->second, std::nullopt);
+            return AddressUFVKMetadata(ufvkId->second, std::nullopt, false);
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/keystore.cpp:499]: Hm, right. I didn't think about the trial-decryption scenario. I think that we should fix this between 4.7.0-rc1 and the release; I'll add a ticket if that's an acceptable approach.

File: src/wallet/wallet.cpp [hunk 19]
```diff
@@ -853,6 +857,123 @@ bool CWallet::LoadUnifiedAddressMetadata(const ZcashdUnifiedAddressMetadata &add
...
+                }
+            },
+            [&](const OrchardRawAddress& addr) {
+                auto ua = self->FindUnifiedAddressByReceiver(addr);
+                if (ua.has_value()) {
+                    return libzcash::PaymentAddress{ua.value()};
+                } else if (ufvk.has_value() && ufvk->GetOrchardKey().has_value()) {
+                    auto orchardKey = ufvk->GetOrchardKey().value();
+                    auto j = orchardKey.ToIncomingViewingKey().DecryptDiversifier(addr);
+                    if (j.has_value()) {
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/wallet.cpp:906]: Yes, this checks that the address matches.

File: src/wallet/wallet.cpp [hunk 20]
```diff
@@ -853,6 +857,123 @@ bool CWallet::LoadUnifiedAddressMetadata(const ZcashdUnifiedAddressMetadata &add
...
+                auto ua = self->FindUnifiedAddressByReceiver(addr);
+                if (ua.has_value()) {
+                    return libzcash::PaymentAddress{ua.value()};
+                } else if (ufvk.has_value() && ufvk->GetSaplingKey().has_value()) {
+                    auto saplingKey = ufvk->GetSaplingKey().value();
+                    auto j = saplingKey.DecryptDiversifier(addr.d);
+                    // std::get is safe here because we know we have a valid Sapling diversifier index
+                    auto defaultUA = std::get<std::pair<UnifiedAddress, diversifier_index_t>>(
+                            ufvk->Address(j, CWallet::DefaultReceiverTypes()));
+                    return libzcash::PaymentAddress{defaultUA.first};
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/wallet.cpp:894]: This also seems like a bug that we can fix between 4.7.0-rc1 and the release.

File: src/keystore.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (str4d) [reply src/keystore.cpp:499]: I'm happy to fix this during RC.

File: src/wallet/wallet.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (str4d) [reply src/wallet/wallet.cpp:894]: I'm happy for this to be fixed during RC.

[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/wallet.cpp:894]: Opened https://github.com/zcash/zcash/issues/5708

File: src/keystore.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/keystore.cpp:499]: Opened https://github.com/zcash/zcash/issues/5708

File: src/rust/include/rust/orchard/wallet.h [hunk 21]
```diff
@@ -263,6 +263,59 @@ void orchard_wallet_get_filtered_notes(
...
+};
+
+/**
+ * A C struct used to transfer Orchard action output information across the FFI boundary.
+ * This must have the same in-memory representation as the `FFIActionOutput` type in
+ * orchard_ffi/wallet.rs.
+ */
+struct RawOrchardActionOutput {
+    uint32_t outputActionIdx;
+    OrchardRawAddressPtr* addr;
```
Comment (ebfull) [inline src/rust/include/rust/orchard/wallet.h:286]: Can this be referred to as "recipient" to be consistent with `FFIActionOutput`?

File: src/rust/include/rust/orchard/wallet.h [hunk 22]
```diff
@@ -263,6 +263,59 @@ void orchard_wallet_get_filtered_notes(
...
+ * actions of that bundle) to the provided result receiver.
+ *
+ * Note that the callbacks can perform any necessary conversion from a
+ * `RawOrchardActionSpend` or `RawOrchardActionOutput` value in addition to modifying the
+ * provided result receiver.
+ *
+ * `raw_ovks` must be a pointer to an array of `unsigned char[32]`.
+ *
+ * The `addr` pointer on each `RawOrchardActionOutput` value must be freed using
+ * `orchard_address_free`.
```
Comment (ebfull) [inline src/rust/include/rust/orchard/wallet.h:308]: What about the `receivedAt` pointer for each `RawOrchardActionSpend`?

File: src/wallet/wallet.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (ebfull) [reply src/wallet/wallet.cpp:906]: I had the same question and checked and indeed it does.

**Review by ebfull — CHANGES_REQUESTED** [review]

**Review by ebfull — APPROVED** [review]

File: src/wallet/wallet.h [hunk 23]
```diff
@@ -1661,16 +1674,16 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
...
     {
         LOCK2(cs_main, cs_wallet);
-        LogPrintf("SaveRecipientMappings:\n%s", txid.ToString());
 
         for (const auto& recipient : recipients)
         {
+            sendRecipients[txid].push_back(recipient);
             if (recipient.ua.has_value()) {
-                CWalletDB(strWalletFile).WriteRecipientMapping(
+                assert(CWalletDB(strWalletFile).WriteRecipientMapping(
```
Comment (daira) [inline src/wallet/wallet.h:1682]: Should this be (logging and) propagating the error rather than asserting?

File: src/wallet/wallet.h [hunk 24]
```diff
@@ -1691,6 +1704,12 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
      */
     static CAmount GetRequiredFee(unsigned int nTxBytes);
 
+    /**
+     * The current set of default receiver types used when the wallet generates
+     * unified addresses
+     */
+    static std::set<libzcash::ReceiverType> DefaultReceiverTypes();
```
Comment (daira) [inline src/wallet/wallet.h:1711]: Shouldn't this be height-dependent?

**Review by daira — COMMENTED** [review]
Post-hoc ACK with comments.

File: src/wallet/wallet.h (hunk 24 above)
Comment (nuttycom) [reply src/wallet/wallet.h:1711]: Yes, eventually.

File: src/wallet/wallet.h (hunk 23 above)
Comment (nuttycom) [reply src/wallet/wallet.h:1682]: Do we consider a database write failure to be a recoverable error? The way that we use `assert` in general is mixed up, as we've discussed in the past, but I think that write failures should generally be catastrophic.

File: src/wallet/wallet.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/wallet.cpp:894]: Fixed in #5715

File: src/rust/include/rust/orchard/wallet.h (hunk 21 above)
Comment (nuttycom) [reply src/rust/include/rust/orchard/wallet.h:286]: Fixed in #5715

File: src/rust/include/rust/orchard/wallet.h (hunk 22 above)
Comment (nuttycom) [reply src/rust/include/rust/orchard/wallet.h:308]: Fixed in #5715

File: src/wallet/wallet.h (hunk 23 above)
Comment (nuttycom) [reply src/wallet/wallet.h:1682]: Fixed in #5715

File: src/wallet/wallet.h (hunk 24 above)
Comment (nuttycom) [reply src/wallet/wallet.h:1711]: Fixed in #5715

File: src/keystore.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/keystore.cpp:499]: Fixed in #5715
